src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -0,0 +1,350 @@
+
+  /** Trait that can be implemented to provide a UMI assignment strategy. */
+  sealed trait UmiAssigner {
+    def assign(rawUmis: Seq[String]) : Map[String, String]
+  }
+
+  /** Assigns UMIs based solely on sequence identity. */
+  class IdentityUmiAssigner extends UmiAssigner {
+    override def assign(rawUmis: Seq[String]): Map[String, String] = rawUmis.map(umi => umi -> umi.toUpperCase).toMap
+  }
+
+  /** Assigns UMIs based solely on sequence identity. */
+  class SimpleErrorUmiAssigner(val maxMismatches: Int) extends UmiAssigner {
+    override def assign(rawUmis: Seq[String]): Map[String, String] = {
+      type UmiSet = mutable.SortedSet[String]
+      val umiSets = ListBuffer[UmiSet]()
+      for (umi <- rawUmis) {
+        val matchedSets = mutable.Set[UmiSet]()
+        for (set <- umiSets) {
+          if (set.exists(other => countMismatches(other, umi) <= maxMismatches)) {
+            set.add(umi)
+            matchedSets.add(set)
+          }
+        }
+
+        matchedSets.size match {
+          case 0 => umiSets += mutable.SortedSet(umi)
+          case 1 => matchedSets.head.add(umi)
+          case _ =>
+            // Merge the sets and remove all but one from umiSets
+            val pick = matchedSets.head
+            matchedSets.tail.foreach(set => {
+              umiSets -= set
+              pick ++= set
+            })
+        }
+      }
+
+      umiSets.flatMap(set => set.map(umi => umi -> set.head)).toMap
+    }
+  }
+
+  /** Class that implements the directed adjacency graph method from umi_tools. */
+  class AdjacencyUmiAssigner(val maxMismatches: Int) extends UmiAssigner {
+    /** Represents a node in the adjacency graph; equality is just by UMI sequence. */
+    private class Node(val umi: String, val count: Long, val children: mutable.Buffer[Node] = mutable.Buffer()) {
+      /** Gets the full set of descendants from this node. */
+      def descendants: List[Node] = {
+        val buffer = ListBuffer[Node]()
+        addDescendants(buffer)
+        buffer.toList
+      }
+
+      private def addDescendants(buffer: mutable.Buffer[Node]): Unit = children.foreach(child => {
+        buffer += child
+        child.addDescendants(buffer)
+      })
+
+      override def equals(other: scala.Any): Boolean = other.isInstanceOf[Node] && this.umi == other.asInstanceOf[Node].umi
+    }
+
+    override def assign(rawUmis: Seq[String]): Map[String, String] = {
+      // Make a list of counts of all UMIs in order from most to least abundant
+      val nodes = SimpleCounter(rawUmis).map(x => new Node(x._1, x._2)).toBuffer.sortBy((n:Node) => -n.count)
+      val roots = mutable.Buffer[Node]()
+
+      // Now build one or more graphs starting with the most abundant remaining umi
+      def addChildren(root: Node, others: mutable.Buffer[Node]) : Unit = {
+        val xs = others.filter(other => countMismatches(root.umi, other.umi) <= maxMismatches && root.count >= 2 * other.count - 1)
```

Comment:
some explanation would be nice here about the intent of what it is doing, especially `root.count >= 2 * other.count - 1`